### ISSUE TITLE
Allow for git-describe-like output if desired

### ISSUE THREAD

**Issue by kworth** [gladed/gradle-android-git-version#67]
Given the following tree

```
*   4a30b81 (feature/foo) Merge branch 'master' into feature/foo    # expected distance: 5
|\
| *   1e79026 (master) Merge branch 'feature/bar'    # distance: 2
| |\
| | * 56f09f5 (feature/bar) Add some more bar feature stuff
| | * fdb2e34 (tag: 0.6.2) Add some bar feature stuff
* | | c62c0ec Add some foo feature stuff
* | | 5a3f533 Merge branch 'master' into feature/foo
|\ \ \
| |/ /
| | /
| * cf1bc69 Make some changes on master
| |
| |
...
```
The versionName generated while on feature/bar (1 commit ahead of tag `0.6.2`) is good and correct as it shows a commit distance of 1 (`0.6.2-1-56f09f5`) just as `git describe` does, and the same is true for versionName while on master (2 commits ahead of tag `0.6.2`), yielding `0.6.2-2-1e79026`.

But, the versionName generated while on `feature/foo` does not match the output of git describe. The commit distance appears to only increment by 1 for the merge commit, whereas `git describe` increments by 1 plus the number of commits being merged in.
```
$ git checkout feature/foo
$ ./gradlew androidgitversion
0.6.2-3-4a30b81
```
The expected output (to get the versionName to match the commit-ish generated by `git describe`) while on feature/foo would be `0.6.2-5-4a30b81`, but the versionName being generated is `0.6.2-3-4a30b81`. It appears to be incrementing by 1 (from the merge commit), but fails to count the commits `c62c0ec` and `5a3f533`.

```
$ git describe feature/foo
0.6.2-5-g4a30b81
```
Thank you.

**Comment by gladed** [comment]
This is currently working as designed, with a merge commit counting as 1 commit no matter how many commits the merge encapsulates. The goal isn't really to match git-describe either - if you need to look up the commit

Is there some reason it is better to have inflated distance counts? Is it just to make `git show` work? (By default there are lots of cases where the output of androidGitVersonName will not match.)

**Comment by kworth** [comment]
Sure. I see lots of power/usefulness in being able to ask a user/tester/fellow developer which version of the app they are using and not have to do any extra step to map that to a commit. I see that as why `git describe` is designed the way it is, and seems natural to me to keep that design/usefulness.

What are the other cases where it won't match? Or rather, if the output of git describe were used to generate the version name, is there anything that would make it additionally challenging to keep them matching?

**Comment by gladed** [comment]
That's a fair use case. I wasn't originally attempting to match `git describe` but I can see how it would make things easier.

Differences I can see from `git-describe` docs and current android-git-version behavior:

* Missing "g" (as you reported in #66, thank you).
* Different commit count (reported here).
* A difference in how the nearest tag is found: git-describe prefers annotated tags over lightweight tags, tags with more recent dates over tags with earlier dates, and lastly tags with fewer commits over tags with more commits as calculated in point 2.

Unfortunately `jgit` doesn't give an easy way to make this calculation although it does provide a `DescribeCommand`. This plugin would have to call that, and try to parse out the rev-count from the result. Then it would match.

**Comment by gladed** [comment]
One point of incompatibility with the current layout is that android-git-version allows you to specify a subtree and only count revs for commits within that subtree. For example if you have a `lib` directory that hasn't experienced any additional commits that change files in that subtree you get the same tag offset count.

Also note that `jgit` itself has had several incompatibilities with git-describe over time. So even if we did use DescribeCommand I can't be 100% sure to have the exact same behavior as `git` itself in all cases.

So it may be safer to just point people to the commit hash if they want to ensure they have a build of the same commit.

**Comment by kworth** [comment]
It may make for a nice feature to allow it to be configured to use git describe. I wish I understood the current codebase better to be able to make a PR, or fork the repo if you don't think it will make the cut.

Let me know what direction you might take things, and I'd be happy to try some changes if you could point me in the right direction for doing so.

**Comment by gladed** [comment]
I don't object to making git-describe-like the behavior the default, but I would want to keep the ability to use format to customize it. I think to implement this one would just re-implement jgit's DescribeCommand behavior, except in Groovy for the plugin, and with sensitivity to only-in.

Of course if you really want git-describe behavior exactly, it's easy to do `name = git.describe(tags: true)` without using the plug-in at all.

**Comment by kworth** [comment]
Not use the plugin? And throw the baby out with the bathwater? No way. The plugin is great, with all its features and flexibility.

So, are these changes something you anticipate working on soon? Or perhaps I can help by getting things started and making a pull request?

**Comment by gladed** [comment]
I would welcome a pull on this!

**Comment by kworth** [comment]
@gladed would you be able to point me in the right direction in how to make and test changes? Is there any sort of shortcut where I can just modify code that's running when I make my Android projects, or will I need to set up this project and learn how to package it for use by my Android projects, etc.? Thanks for any pointers you can provide while I start googling to understand all this stuff better.

**Comment by gladed** [comment]
Really the best way is unit test. Each unit test starts with an empty repo to which you can commit changes, apply tags, branch, merge etc.

After that you can publish to a local maven repo and try from a sample project if you like.

**Comment by kworth** [comment]
@gladed let me know what you think about PR #68 just as an initial/general direction.

**Comment by kworth** [comment]
@gladed What if instead of trying to find "the best rev count", we could simply define two different rev counts that have different meanings and purposes? The existing count being used by the plugin currently could be the "minimum count", or the "single parent count", or... I don't know. And then the count which is provided by `git describe` could be the "total count", or "comprehensive count". Or maybe one is "count" and the other is "distance", or something.

If we did something like that, then the user could customize their format as they please. The code could use `git describe` as its initial base, and then the formatting code could do its thing. As an example - for me personally - I simply want my format to be git_describe-branch-dirty and hide master. But given that format uses a commit count/distance that others don't want, they could simply use a configuration to indicate they want the *other* count.

Curious to hear your thoughts, and I hope that makes sense.

**Comment by gladed** [comment]
Released in 0.4.12
